### PR TITLE
Model IO in JSON.

### PR DESCRIPTION
**Pull request by trivialfis**
The "model" part of https://github.com/dmlc/xgboost/pull/4732.  Follow up PRs will be for "Configuration".

### CODE REVIEW

**Comment by trivialfis** [comment]
@hcho3 The schema of tree is different from our original design, suggested by @trams I split up each components of a node for future extensibility.  Will update the schema accordingly.

File: src/c_api/c_api.cc [hunk 1]
```diff
@@ -567,23 +569,43 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
...
+    fo->Write(str.c_str(), str.size());
+  } else {
+    auto *bst = static_cast<Learner*>(handle);
+    bst->Save(fo.get());
+  }
   API_END();
 }
 
+// The following two functions are `Load` and `Save` for memory based serialization
+// methods. E.g. Python pickle.
```
Comment (hcho3) [inline src/c_api/c_api.cc]: Note. We may consider using JSON for Python pickling in the future.

Comment (trivialfis) [reply src/c_api/c_api.cc]: @hcho3 See documents in : https://github.com/dmlc/xgboost/pull/4732  Specifically the `saving_model.rst` file.

Comment (hcho3) [reply src/c_api/c_api.cc]: @trivialfis So `enable_experimental_json_serialization` lets you use JSON for pickling. Thanks!

File: src/learner.cc [hunk 2]
```diff
@@ -266,14 +266,61 @@ class LearnerImpl : public Learner {
...
+    learner["learner_model_param"] = mparam_.ToJson();
+    learner["gradient_booster"] = Object();
+    auto& gradient_booster = learner["gradient_booster"];
+    gbm_->SaveModel(&gradient_booster);
+
+    learner["objective"] = Object();
+    auto& objective_fn = learner["objective"];
+    obj_->SaveConfig(&objective_fn);
+
+    learner["attributes"] = Object();
```
Comment (hcho3) [inline src/learner.cc]: @trivialfis I vaguely recall that you wanted to remove extra attributes field. Are we keeping it now?

Comment (trivialfis) [reply src/learner.cc]: @hcho3 Yup.  It saves stuffs like `best_score` and other similar stuffs.  I don't want to add more changes.

File: tests/cpp/gbm/test_gblinear.cc [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  model = model["model"];
+
+  {
+    auto weights = get<Array>(model["weights"]);
+    ASSERT_EQ(weights.size(), 17);
+  }
+
+  {
+    model = Json::Load({model_str.c_str(), model_str.size()});
+    model = model["model"];
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/gbm/test_gblinear.cc]: @hcho3 It involves a deep copy.  Use reference if you don't want a copy

File: tests/cpp/gbm/test_gblinear.cc [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    auto weights = get<Array>(model["weights"]);
+    ASSERT_EQ(weights.size(), 17);  // 16 + 1 (bias)
+  }
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/gbm/test_gblinear.cc]: @hcho3 No, the model is reloaded.  But I will structure it more clearly.

File: tests/cpp/test_learner.cc [hunk 5]
```diff
@@ -112,83 +114,54 @@ TEST(Learner, Configuration) {
...
   {
-    // Create a scope to close the stream before next read.
-    std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(fname.c_str(), "w"));
-    learner->Save(fo.get());
+    std::unique_ptr<Learner> learner { Learner::Create({p_dmat}) };
+    learner->SetParam("verbosity", "3");
+    for (int32_t iter = 0; iter < kIters; ++iter) {
+      learner->UpdateOneIter(iter, p_dmat.get());
+    }
+    learner->SetAttr("bset_score", "15.2");
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/test_learner.cc]: @hcho3 Typo, opening a new PR for addressing your comments.

File: tests/cpp/gbm/test_gblinear.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply tests/cpp/gbm/test_gblinear.cc]: So the model is loaded from `model_str` twice?

File: tests/cpp/test_learner.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/test_learner.cc]: @hcho3 See https://github.com/dmlc/xgboost/pull/5112 .

**Pull request merged**